src/pybind/mgr/rook/rook_cluster.py

```diff
@@ -473,15 +472,164 @@ def filter_devices(self, rook_pods: KubernetesResource, drive_group: DriveGroupS
+
+        self.osd_ids = osd_ids
+        self.replace_flag = replace_flag
+        self.force_flag = force_flag
+
+        self.mon_command = mon_command
+
+        self.patch = patch
+        self.rook_env = rook_env
+
+        self.pods = pods
+        self.inventory = inventory
+
+        self.jobs: KubernetesResource = KubernetesResource(self.batchV1_api.list_namespaced_job, namespace='rook-ceph')
+        self.pvcs: KubernetesResource = KubernetesResource(self.coreV1_api.list_namespaced_persistent_volume_claim, namespace='rook-ceph')
+
+
+    def remove_device_sets(self) -> str:
+        self.to_remove: Dict[str, int] = {}
+        self.pvc_to_remove: List[str] = []
+        for pod in self.pods.items:
+            if (
+                hasattr(pod, 'metadata') 
+                and hasattr(pod.metadata, 'labels') 
+                and 'osd' in pod.metadata.labels 
+                and pod.metadata.labels['osd'] in self.osd_ids
+            ):
+                if pod.metadata.labels['ceph.rook.io/DeviceSet'] in self.to_remove:
+                    self.to_remove[pod.metadata.labels['ceph.rook.io/DeviceSet']] = self.to_remove[pod.metadata.labels['ceph.rook.io/DeviceSet']] + 1
+                else:
+                    self.to_remove[pod.metadata.labels['ceph.rook.io/DeviceSet']] = 1
+                self.pvc_to_remove.append(pod.metadata.labels['ceph.rook.io/pvc'])
+        def _remove_osds(current_cluster, new_cluster):
+            # type: (ccl.CephCluster, ccl.CephCluster) -> ccl.CephCluster
+            assert new_cluster.spec.storage is not None and new_cluster.spec.storage.storageClassDeviceSets is not None
+            for _set in new_cluster.spec.storage.storageClassDeviceSets:
+                    if _set.name in self.to_remove:
+                        if _set.count == self.to_remove[_set.name]:
+                            new_cluster.spec.storage.storageClassDeviceSets.remove(_set)
+                        else:
+                            _set.count = _set.count - self.to_remove[_set.name]
+            return new_cluster
+        return self.patch(ccl.CephCluster, 'cephclusters', self.rook_env.cluster_name, _remove_osds)
+
+    def check_force(self) -> None:
+        if not self.force_flag:
+            safe_args = {'prefix': 'osd safe-to-destroy',
+                        'ids': [str(x) for x in self.osd_ids]}
+            ret, out, err = self.mon_command(safe_args)
+            if ret != 0:
+                log.error(f"error when checking if osds are safe to destroy, {err}")
+                raise RuntimeError(err)
+
+    def set_osds_down(self) -> None:
+        down_flag_args = {
+            'prefix': 'osd down',
+            'ids': [str(x) for x in self.osd_ids]
+        }
+        ret, out, err = self.mon_command(down_flag_args)
+        if ret != 0:
+            log.error(f"error setting down flags on OSDs, {err}")
+            raise RuntimeError(err)
+
+    def scale_deployments(self) -> None:
```

Comment:
Testing it fails because i have no rights to scale deployments .. and i think that also no rights to delete  deployments,.. 

```
RuntimeError: Remote method threw exception: Traceback (most recent call last):
  File "/usr/share/ceph/mgr/rook/module.py", line 484, in remove_osds
    res = self._rook_cluster.remove_osds(osd_ids, replace, force, self.mon_command)
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 1031, in remove_osds
    return self.remover.remove()
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 611, in remove
    self.scale_deployments()
  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 560, in scale_deployments
    replicas=0
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api/apps_v1_api.py", line 3936, in patch_namespaced_deployment_scale
    (data) = self.patch_namespaced_deployment_scale_with_http_info(name, namespace, body, **kwargs)  # noqa: E501
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api/apps_v1_api.py", line 4041, in patch_namespaced_deployment_scale_with_http_info
    collection_formats=collection_formats)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 345, in call_api
    _preload_content, _request_timeout)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 176, in __call_api
    _request_timeout=_request_timeout)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/api_client.py", line 404, in request
    body=body)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/rest.py", line 298, in PATCH
    body=body)
  File "/usr/lib/python3.6/site-packages/kubernetes/client/rest.py", line 231, in request
    raise ApiException(http_resp=r)
kubernetes.client.rest.ApiException: (403)
Reason: Forbidden
```

Reply:
[This PR](https://github.com/rook/rook/pull/8568) should've fixed that